From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 20 Sep 2024 08:20:11 +0300
Subject: [PATCH] Extra recipes API

Allows to set whether the recipe is special.
Special recipes do not get sent to the client and bypass the limited crafting game rule.

diff --git a/src/main/java/org/bukkit/inventory/CookingRecipe.java b/src/main/java/org/bukkit/inventory/CookingRecipe.java
index 07906ca1a9b39fcc6774870daa498402f7f37917..08a386efd66044fa6b093f37ba544b3e80bbcd4f 100644
--- a/src/main/java/org/bukkit/inventory/CookingRecipe.java
+++ b/src/main/java/org/bukkit/inventory/CookingRecipe.java
@@ -20,6 +20,7 @@ public abstract class CookingRecipe<T extends CookingRecipe> implements Recipe,
     private int cookingTime;
     private String group = "";
     private CookingBookCategory category = CookingBookCategory.MISC;
+    private boolean special = false; // Kiterino - Extra recipes API
 
     /**
      * Create a cooking recipe to craft the specified ItemStack.
@@ -198,4 +199,27 @@ public abstract class CookingRecipe<T extends CookingRecipe> implements Recipe,
         Preconditions.checkArgument(category != null, "category cannot be null");
         this.category = category;
     }
+
+    // Kiterino start - Extra recipes API
+    /**
+     * Checks whether this recipe is special.
+     * <br>Special recipes do not get sent to the client
+     * and bypass the limited crafting game rule.
+     *
+     * @return whether this recipe is special
+     */
+    public boolean isSpecial() {
+        return this.special;
+    }
+
+    /**
+     * Sets whether this recipe is special
+     *
+     * @param special whether this recipe is special
+     * @see #isSpecial()
+     */
+    public void setSpecial(boolean special) {
+        this.special = special;
+    }
+    // Kiterino end - Extra recipes API
 }
diff --git a/src/main/java/org/bukkit/inventory/CraftingRecipe.java b/src/main/java/org/bukkit/inventory/CraftingRecipe.java
index 5bf55b40fbf6ec708f37d90bd0853fe7dd8fffd9..87d9b2acc66d9854d00feab09bb9ae7a01ed7f2d 100644
--- a/src/main/java/org/bukkit/inventory/CraftingRecipe.java
+++ b/src/main/java/org/bukkit/inventory/CraftingRecipe.java
@@ -16,6 +16,7 @@ public abstract class CraftingRecipe implements Recipe, Keyed {
     private final ItemStack output;
     private String group = "";
     private CraftingBookCategory category = CraftingBookCategory.MISC;
+    private boolean special = false; // Kiterino - Extra recipes API
 
     protected CraftingRecipe(@NotNull NamespacedKey key, @NotNull ItemStack result) {
         Preconditions.checkArgument(key != null, "key cannot be null");
@@ -102,4 +103,27 @@ public abstract class CraftingRecipe implements Recipe, Keyed {
         Preconditions.checkArgument(!result.isEmpty(), "Recipe cannot have an empty result."); // Paper
         return result;
     }
+
+    // Kiterino start - Extra recipes API
+    /**
+     * Checks whether this recipe is special.
+     * <br>Special recipes do not get sent to the client
+     * and bypass the limited crafting game rule.
+     *
+     * @return whether this recipe is special
+     */
+    public boolean isSpecial() {
+        return this.special;
+    }
+
+    /**
+     * Sets whether this recipe is special
+     *
+     * @param special whether this recipe is special
+     * @see #isSpecial()
+     */
+    public void setSpecial(boolean special) {
+        this.special = special;
+    }
+    // Kiterino end - Extra recipes API
 }
